DNA.Prjct1.iPadchenko

YuryRadchenko opened the pull request.

YuryRadchenko committed:
Проект 1. ДНК.

Первая работа по созданию ДНК, нахождению между ними различий и мутации.

YuryRadchenko committed:
Правки 1.1: проверка повторения

Изменения в - (void) mutate: (int) xPercentMutator {
 Добавлена проверка: действительно ли ячейка будет меняться.

YuryRadchenko committed:
Оптимизация и косметические изменения в коде

- В процедуре мутации использовал NSMutableSet
- Косметические изменения в коде

YuryRadchenko committed:
Надеюсь что финальная версия

Внес проверку введенных процентов ниже 0. Раньше была проверка только
выше 100%.

oluckyman commented:
У меня не компилится - говорит что проперти DNA не объявлено в имплементации.
Наверно у нас разные XCode и они по-разному со свойствами работают... У меня 4.3.2.

oluckyman commented on dnk.Prj1.iPadchenko/dnk.Prj1.iPadchenko/Cell.m:
```diff
@@ -0,0 +1,76 @@
+
+- (id) init {
+    
+    self = [super init];//вызываем инициализация от класса родителя, считаем это инициализацией себя
```
Камменты такого типа бессмысленные. Они требуют времени чтобы их прочесть, но не несут в себе полезной информации

YuryRadchenko replied:
согласен. со временем набью руку и перестану писать. пока пишу больше для себя.

oluckyman commented on dnk.Prj1.iPadchenko/dnk.Prj1.iPadchenko/Cell.m:
```diff
@@ -0,0 +1,76 @@
+        for (int i=0; i < dnaLenght; i++)
+            
+            [self.DNA addObject:[self randomCellValueATGC]]; //добавляем в конец массива случайное значение
+        
```
Бросается в глаза небрежность  кода и несоблюдение правил хорошего тона. Например, блоки циклов хорошо обёртывать в фигурные скобки, даже если блок из одной строки.

YuryRadchenko replied:
Скобки умышленно удалил. Не знал что это по этикету положено. Спасибо, буду знать.

oluckyman replied:
тут нет "положено" и "не положено". Так же, как нет правила "клади грабли зубцами вниз".  Эти "правила" формируются из опыта.  
Блоки без скобок -- потенциальные источники ошибок. Особенно, если это чужой код. Лучше сразу привыкнуть быть педантом в таких мелочах, чтоб потом не наступать на грабли

oluckyman commented on dnk.Prj1.iPadchenko/dnk.Prj1.iPadchenko/Cell.m:
```diff
@@ -0,0 +1,76 @@
+            hD++; //учитываем несовпадение
+    
+    return hD;
+}
```
Зачем после каждой строки идёт пустая строка?

YuryRadchenko replied:
Мне так читабельней.

oluckyman commented on dnk.Prj1.iPadchenko/dnk.Prj1.iPadchenko/Cell.m:
```diff
@@ -0,0 +1,76 @@
+    
+    //NSLog(@"name = %@", self);//расскоментировать, если надо видеть класс и адрес переменной
+    NSLog(@"ДНК = \n%@",dnaOneString);
+
```
Тут можно было обойтись без этого цикла. Есть метод, который из элементов массива склеивает строку. К сожалению, забыл как назвается.

hmmvot replied:
``` objective-c
[DNA componentsJoinedByString: @","];
```

YuryRadchenko replied:
Понял, спасибо.

oluckyman commented on dnk.Prj1.iPadchenko/dnk.Prj1.iPadchenko/Cell.m:
```diff
@@ -0,0 +1,76 @@
+        case 2: ATGC = @"G"; break;
+        case 3: ATGC = @"C"; break;
+    }
+    
```
Более "программистый" подход в данном случае - завести массив с элементами A T G C и по рандомному индексу выбирать один из его элементов.

YuryRadchenko replied:
Возможно. Я исходил из того, что более чем АTGC не будет, поэтому массив решил не трогать.

oluckyman commented on dnk.Prj1.iPadchenko/dnk.Prj1.iPadchenko/main.m:
```diff
@@ -0,0 +1,104 @@
+#import "Cell.h"
+
+@interface Cell (mutator)
+
```
в XCode можно добавить категорию через меню файл. Так же как новый класс. Тогда хкод всю грязную работу делает сам - создаёт отдельные файлы .h и .m где хранит описание новой категории.
Добавлять категорию в файл main.m неправильно.

YuryRadchenko replied:
В постановке задачи указано "6. Создайте категорию mutator класса Cell. ..." . Или я не о том?

hmmvot replied:
Вынесите категорию из main.m в отдельные файлы.

YuryRadchenko replied:
Я все еще не понимаю насчет категорий. В постановке задаче говорится сделать именно так. Что я не так сделал?

hmmvot replied:
У вас все так. Но категорию лучше описать в отдельном файле (так, как вы поступили с классом Cell, например).

YuryRadchenko replied:
Я думал, у нас практика повторяет рассказанный материал в обучении. Например этот http://www.youtube.com/watch?v=6cAO032M6Yk&feature=player_detailpage#t=46s

oluckyman commented on dnk.Prj1.iPadchenko/dnk.Prj1.iPadchenko/main.m:
```diff
@@ -0,0 +1,104 @@
+                
+                //запоминаем выбор
+                [cellMutator addObject:[NSString stringWithFormat:@"%d", randomIndex]];
+                
```
для хранения целых числе в множестве больше подошёл бы класс NSNumber

YuryRadchenko replied:
Добро, поправлю.

YuryRadchenko replied:
Решил не менять типа массива на NSNumber. В это классе надо будет прибегать к циклу проверки каждого идекса массива. Видимо там нельзя получить аналогию строки:
[cellMutator containsObject:[NSString stringWithFormat:@"%d", randomIndex]]

hmmvot replied:
Что мешает использовать containsObject: у массива наполненного NSNumber?

hmmvot commented on dnk.Prj1.iPadchenko/dnk.Prj1.iPadchenko/Cell.m:
```diff
@@ -0,0 +1,76 @@
+}
+
+
+- (void) print {
```
Для такого уже существует метод description, перегрузите его.

hmmvot replied:
Возможно вам пригодиться такой конструктор NSString:

``` objective-c
[NSString stringWithFormat: @"ДНК = %@", [self.DNA componentsJoinedByString: @", "]];
```

YuryRadchenko replied:
Да, спасибо. Упрощу.

hmmvot commented on dnk.Prj1.iPadchenko/dnk.Prj1.iPadchenko/main.m:
```diff
@@ -0,0 +1,104 @@
+    
+    else
+        
+        NSLog(@"Некорректно введен процент мутации");
```
Если вы хотите обрабатывать ошибки то пусть ваш метод mutate возвращает YES, если все ок, и NO в противном случае.
В этой функции не должно быть вывода в лог.

YuryRadchenko replied:
Интересная мысль. Так и сделаю. Спасибо.

YuryRadchenko replied:
> пусть ваш метод mutate возвращает YES, если все ок, и NO в противном случае.
> Это противоречит постановке задачи: "п.6. ...  В ней опишите метод mutate, который возвращает void и принимает int."

hmmvot replied:
Тогда уроните программу с помощью @throw или assert (NSAssert).

YuryRadchenko commented:
У меня XCode Version 4.5.1 (4G1004) . Можно без @property . У меня все компилируется и выполняется. Кажется в видео об этом говорили.

YuryRadchenko committed:
Доработки с учетом некоторых замечаний

С учетом замечаний и моих доработок:
- Сокращение количества строк в процедуре Печати
- Избавился от логической переменной в процедуре мутации
- Использовал массив для ATGC
- добавил {} у if где должен выполняться только 1 оператор
